### DISCUSSION TITLE
Unsure how to deal with postgres bpchar

### DISCUSSION THREAD

**Discussion by valhuber** [sqlalchemy/sqlalchemy#8592] [Usage Questions]
# Test Description

In [this repo](https://github.com/valhuber/postgres_bpchar), I have a postgres (PostgreSQL) 13.4 (Debian 13.4-1.pgdg100+1) database, running on Mac, under docker:

```
docker run -d --name postgresql-container --net dev-network -p 5432:5432 -e PGDATA=/pgdata -e POSTGRES_PASSWORD=p apilogicserver/postgres:version1.0.2
```

The following table uses `bpchar` for a datatype:

<figure><img src="https://github.com/valhuber/postgres_bpchar/blob/main/images/table-def.png?raw=true"></figure>

When running the `reflect` code under `SQLAlchemy==1.4.29`, I get the following console log:

```
/Users/val/dev/examples/postgres_bpchar/run.py:7: SAWarning: Did not recognize type 'bpchar' of column 'customer_id'
```

The metadata sees this as `nulltype`:

<figure><img src="https://github.com/valhuber/postgres_bpchar/blob/main/images/null-type.png?raw=true"></figure>

# Questions

My questions are:

1. Is `bpchar` supported in the postgres dialect?
   * I thought that the dialect was inferred from the uri `postgresql`.
   * I _no such module_ with `postgres+psycopg2`
2. If not, how can I recover the original type?

**Comment by zzzeek** [comment]
> 
>     1. Is `bpchar` supported in the postgres dialect?

if you have a table with a bpchar column, it should work fine, as the driver in use will generally send string data to and from values of this datatype without SQLAlchemy really having to know much about it.     the SQLAlchemy dialect itself doesn't have the name "bpchar" in its lookup table so it doesnt know how to make a decision here on how to reflect this, but the Table object you get back will probably do everything its supposed to without much problem.

>       
>        * I thought that the dialect was inferred from the uri `postgresql`.

it is

>        * I _no such module_ with `postgres+psycopg2`

"postgresql" - the name "postgres" was deprecated 11 years ago and was removed in the past year.


> 
>     2. If not, how can I recover the original type?

assuming "bpchar" is just a String you could take the String datatype and store it in the dialect level [ischema_names](https://github.com/sqlalchemy/sqlalchemy/blob/7608e2f04806b767f1f7f8717d4d665537e0c8f6/lib/sqlalchemy/dialects/postgresql/base.py#L1520) dictionary, but if "bpchar" is some datatype that is expected to be common in user-defined table schemas we'd probably want to add a type for it.

that said, "bpchar" does not seem to be a commonly used type and it's [not listed](https://www.postgresql.org/docs/14/datatype-character.html) in their public doc for datatypes, it seems to be referred towards as the [internal name for the character datatype](https://www.postgresql.org/docs/current/typeconv-query.html).  while we should likely add a placeholder for "bpchar" as we have other such internal placeholders present, it's a little odd for there to be a Table made with "bpchar" explicitly, it could just use "char".

**Reply by valhuber** [reply]
Thanks for the prompt reply.

I am aware it "kinda works" for normal access, but the SAFRS API package checks data types and becomes concerned on unrecognized types, particularly in keys.

So, I tried your suggestion in ischema names, wanted to confirm this is what you meant:

```
from sqlalchemy import CHAR, Column, DateTime, Float, ForeignKey, Index, Integer, String, TIMESTAMP, Table, Text, UniqueConstraint, text

# db: docker run -d --name postgresql-container --net dev-network -p 5432:5432 -e PGDATA=/pgdata -e POSTGRES_PASSWORD=p apilogicserver/postgres:version1.0.2

substitute_char = True
if substitute_char:
    pg_dialect.ischema_names["bpchar"] =  CHAR
```

It does work... I was wondering if there is a **_more general approach_** that is not db-specific.  Eg, I'd like API Logic Server to accept a datatype_map.yaml, which provides mappings for various databases, which the user can extend if the need arises.  This is not for "new" datatypes, just to map odd ones to the standard counterparts...

BTW, got the db [from here](https://github.com/pthom/northwind_psql/blob/master/northwind.sql).

**Reply by zzzeek** [reply]
> It does work... I was wondering if there is a **_more general approach_** that is not db-specific. Eg, I'd like API Logic Server to accept a datatype_map.yaml, which provides mappings for various databases, which the user can extend if the need arises. This is not for "new" datatypes, just to map odd ones to the standard counterparts...

well you can add that entry to the PG dialect regardless of what database runs.     We also, as I mentioned, can just add it.

however, 

> 
> BTW, got the db [from here](https://github.com/pthom/northwind_psql/blob/master/northwind.sql).

I would suggest this is really a bug in this particular "northwind" sample.  The northwind sample is supposed to be an extremely traditional database sample used for learning.  however they are relying upon an *extremely* obscure, internal-use PostgreSQL datatype (I've been googling this thing for 30 minutes and it makes absolutely not sense what it is - how can you "blank pad" a char field that **has no length**?   all the announcements that it's "the same as char()" are wrong, in that **there is no unlengthed option** for the `char` type in PostgreSQL, the length defaults to 1 if you don't specify it.)

The use of this type is wrong on the part of that database quite frankly, this is an undocumented internal structure of PostgreSQL and doesn't belong in a publicly consumable schema.

**Comment by valhuber** [comment]
I am with you there… particularly the BP with no length?  This developer is clearly relying on the kindness of strangers…

And I googled all morning ;)

**Reply by zzzeek** [reply]
I think what "bpchar" would mean is that if you run some SQL expression on a char field that is blank padded, you get back the "bpchar" datatype which PG knows the blank spaces that are present, regardless of length, are padding , and not part of the content.     But that doesnt make any sense as a fixed table datatype.

**Comment by valhuber** [comment]
I believe this is actually a practice that was common in the mainframe days… think, punched card (!).

**Reply by zzzeek** [reply]
well sure, it's for fixed length data, that's extremely common. but there has to be a length for that to work.

**Comment by pthom** [comment]
I just pushed a [commit](https://github.com/pthom/northwind_psql/commit/4aa7967a33b84e693db56d4c23ada410a3ff04a6) that corrects this on the northwind example

**Reply by zzzeek** [reply]
@pthom thanks so much!!  PG should also document this datatype better since you can use it in CREATE TABLE.
